### PR TITLE
Dates: fix bugs and add tests

### LINKED ISSUE #88: [PE-D] Invalid date values are not handled
When invalid month or date values are used. There is no error message thrown or any indication that the input is invalid. The app just remains the same and users will not know what is wrong and what is happening.

[Screen Recording 2021-10-29 at 16.24.29.mov](https://raw.githubusercontent.com/sheehui/ped/main/files/4691a2cf-a0ec-4685-8195-ffb5756a327b.mov)

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: sheehui/ped#1

### LINKED ISSUE #92: [PE-D] Does not indicate today for dates when monthly/yearly is used
Seems quite odd that the app shows how a date is "in xx days" but does not show that it is happening today. I think users will want to know that someones birthday is today....and find it quite important actually.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: sheehui/ped#2

### PR DESCRIPTION
**Pull request by bnjmnt4n**
- Strings which match the date format but are not actual dates (eg.`2020-02-32`) should not be recognized as valid dates.
  Closes #88, #97, #105.

- A bug in the logic caused the next occurence of recurring dates to always be the date in the future when the reference date was the same date as the recurring date.
  Closes #92.

### CODE REVIEW

File: src/main/java/socialite/model/person/Date.java [hunk 1]
```diff
@@ -14,21 +16,22 @@
...
      * Construct a new {@code Date} with a given name and value,
      * by parsing the input string.
      * @param date The input string to parse.
      */
     public Date(String date) {
-        Pattern pattern = Pattern.compile(VALIDATION_REGEX);
-        Matcher matcher = pattern.matcher(date);
+        Matcher matcher = VALIDATION_PATTERN.matcher(date);
+        boolean matches = matcher.matches();
+        assert matches;
```
Comment (david-eom) [inline src/main/java/socialite/model/person/Date.java:33]: Shiet defensive programming meister

**Review by david-eom — APPROVED** [review]
Thanks Ben for fixing the bug! Will give the approval first, but may I trouble you to impose a character limit for the description of the date, something like #149? This is to resolve #114. Other than that LGTM!

Current situation when user is malicious:
截圖 2021-11-05 下午4 34 56

File: src/main/java/socialite/model/person/Date.java [hunk 2]
```diff
@@ -66,32 +69,62 @@ public LocalDate getDate() {
         }
 
         Period interval = getIntervalPeriod(recurrenceInterval);
+        assert interval != null;
```
Comment (Zacchua) [inline src/main/java/socialite/model/person/Date.java:72]: Shiet defensive programming meister

**Review by Zacchua — APPROVED** [review]
LGTM👍 thanks Ben!

**Comment by bnjmnt4n** [comment]
I think it would be better to merge this in first, and I'll work on the UI fix in another PR. Thanks for the reviews!

**Pull request merged**